cmd/erasure-server-pool-decom.go

```diff
@@ -0,0 +1,413 @@
+	}
+	return false
+}
+
+func (p poolMeta) Suspend(idx int) bool {
+	if p.IsSuspended(idx) {
+		return false
+	}
+	p.Pools[idx].Suspend = true
+	p.Pools[idx].LastUpdate = time.Now()
+	return true
+}
+
+func (p poolMeta) IsSuspended(idx int) bool {
+	return p.Pools[idx].Suspend
+}
+
+func (p *poolMeta) load(ctx context.Context, set *erasureSets, npools int) (bool, error) {
+	gr, err := set.GetObjectNInfo(ctx, minioMetaBucket, poolMetaName,
+		nil, http.Header{}, readLock, ObjectOptions{})
+	if err != nil && !isErrObjectNotFound(err) {
+		return false, err
+	}
+	if isErrObjectNotFound(err) {
+		return true, nil
+	}
+	defer gr.Close()
+
+	if err = p.DecodeMsg(msgp.NewReader(gr)); err != nil {
+		return false, err
+	}
+
+	switch p.Version {
+	case poolMetaV1:
+	default:
+		return false, fmt.Errorf("unexpected pool meta version: %s", p.Version)
+	}
+
+	// Total pools cannot reduce upon restart
+	if len(p.Pools) > npools {
+		return false, fmt.Errorf("unexpected number of pools provided expecting %d, found %d - please check your command line", len(p.Pools), npools)
+	}
+
+	return len(p.Pools) != npools, nil
+}
+
+func (p poolMeta) Clone() poolMeta {
+	meta := poolMeta{
+		Version: p.Version,
+	}
+	meta.Pools = append(meta.Pools, p.Pools...)
+	return meta
+}
+
+func (p poolMeta) save(ctx context.Context, set *erasureSets) error {
+	buf, err := p.MarshalMsg(nil)
+	if err != nil {
+		return err
+	}
+	br := bytes.NewReader(buf)
+	r, err := hash.NewReader(br, br.Size(), "", "", br.Size())
+	if err != nil {
+		return err
+	}
+	_, err = set.PutObject(ctx, minioMetaBucket, poolMetaName,
+		NewPutObjReader(r), ObjectOptions{})
+	return err
+}
+
+const (
+	poolMetaName = "pool.meta"
+	poolMetaV1   = "1"
+)
+
+// Init() initializes pools and saves additional information about them
+// in pool.meta, that is eventually used for draining the pool, suspend
+// and resume.
+func (z *erasureServerPools) Init(ctx context.Context) error {
+	meta := poolMeta{}
+
+	update, err := meta.load(ctx, z.serverPools[0], len(z.serverPools))
+	if err != nil {
+		return err
+	}
+
+	// if no update is needed return right away.
+	if !update {
+		z.poolMeta = meta
+		return nil
+	}
+
+	meta = poolMeta{}
+
+	// looks like new pool was added we need to update,
+	// or this is a fresh installation (or an existing
+	// installation)
+	meta.Version = "1"
+	for idx := range z.serverPools {
+		meta.Pools = append(meta.Pools, PoolInfo{
+			ID:         idx,
+			Suspend:    false,
+			LastUpdate: time.Now(),
+		})
+	}
+	if err = meta.save(ctx, z.serverPools[0]); err != nil {
+		return err
+	}
+	z.poolMeta = meta
+	return nil
+}
+
+func (z *erasureServerPools) drainObject(ctx context.Context, bucket, object string, gr *GetObjectReader) (err error) {
+	defer gr.Close()
+	if gr.ObjInfo.Multipart {
+		uploadID, err := z.NewMultipartUpload(ctx, bucket, object,
+			ObjectOptions{
+				VersionID:   gr.ObjInfo.VersionID,
+				MTime:       gr.ObjInfo.ModTime,
+				UserDefined: gr.ObjInfo.UserDefined,
```

Comment:
Internal metadata used by tiering such as transition tier name, object name etc are not part of this map. Version Purge status used by replication is also not part of this metadata - instead these fields are part of ObjectInfo and will need to be preserved

Reply:
Will be fixed @poornas it would be helpful to point me to exact metadata key names that would help in setting them back on the objects during draining. 

Transitioning I need to add some other special behavior to handle it, but would you mind giving me a list for replication metadata?